File: app/src/main/java/com/ddancn/keepaccount/fragment/AddFragment.java
package com.ddancn.keepaccount.fragment;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.RadioGroup;
import android.widget.Spinner;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.blankj.utilcode.util.ToastUtils;
import com.ddancn.keepaccount.R;
import com.ddancn.keepaccount.activity.SettingActivity;
import com.ddancn.keepaccount.activity.UpdateActivity;
import com.ddancn.keepaccount.constant.TypeEnum;
import com.ddancn.keepaccount.dao.RecordDao;
import com.ddancn.keepaccount.dao.TypeDao;
import com.ddancn.keepaccount.entity.Record;
import com.ddancn.keepaccount.entity.Type;
import com.ddancn.lib.base.BaseFragment;
import com.ddancn.lib.util.DateUtil;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

/**
 * @author ddan.zhuang
 * 添加或修改记录
 */
public class AddFragment extends BaseFragment {

    private DatePicker datePicker;
    private EditText etMoney;
    private EditText etDetail;
    private RadioGroup rgType;
    private Spinner spinnerType;
    private Button btnAdd;

    private int type = TypeEnum.OUT.value();
    private boolean isUpdate = false;
    private Record recordToUpdate = new Record();

    @Override
    protected int bindLayout() {
        return R.layout.fragment_add;
    }

    @Override
    protected String setHeaderTitle() {
        return getString(R.string.app_name);
    }

    @Override
    protected boolean hasHeader() {
        return !isUpdate;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (context instanceof UpdateActivity) {
            isUpdate = true;
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getActivity() instanceof UpdateActivity) {
            recordToUpdate = ((UpdateActivity) getActivity()).getRecordToUpdate();
        }
    }

    @Override
    protected void initView(View root) {
        setRightImage(R.drawable.ic_setting);
        datePicker = root.findViewById(R.id.date_picker);
        etMoney = root.findViewById(R.id.et_money);
        etDetail = root.findViewById(R.id.et_detail);
        rgType = root.findViewById(R.id.rg_type);
        spinnerType = root.findViewById(R.id.spinner_type);
        btnAdd = root.findViewById(R.id.btn_add);
    }

    @Override
    protected void bindListener() {
        setRightClickListener(v -> startActivity(new Intent(getContext(), SettingActivity.class)));
        // 获取radio group收支选项，联动改变类型spinner的内容
        rgType.setOnCheckedChangeListener((group, checkId) -> {
            type = checkId == R.id.rb_type_out ? TypeEnum.OUT.value() : TypeEnum.IN.value();
            setSpinnerContent();
        });

        // 按钮点击事件
        btnAdd.setOnClickListener(v -> {
            if (!checkParamLegal()) {
                return;
            }
            if (saveOrUpdateRecord()) {
                ToastUtils.showShort(isUpdate ? R.string.update_succeed : R.string.add_succeed);
                if (isUpdate && getActivity() != null) {
                    getActivity().finish();
                } else {
                    reset();
                }
            } else {
                ToastUtils.showShort(isUpdate ? R.string.update_fail : R.string.add_fail);
            }
        });
    }

    /**
     * 判断是否合法
     *
     * @return 是否合法
     */
    private boolean checkParamLegal() {
        if (TextUtils.isEmpty(etMoney.getText().toString())) {
            ToastUtils.showShort(R.string.add_need_money);
            return false;
        }
        if (TextUtils.isEmpty((String) spinnerType.getSelectedItem())) {
            ToastUtils.showShort(R.string.add_need_type);
            return false;
        }
        return true;
    }

    /**
     * 修改页面，先填上原来的数据
     */
    private void initData() {
        btnAdd.setText(getString(R.string.update_btn));
        String[] date = recordToUpdate.getDate().split("-");
        int year = Integer.parseInt(date[0]);
        int month = Integer.parseInt(date[1]);
        int dayOfMonth = Integer.parseInt(date[2]);
        datePicker.updateDate(year, month - 1, dayOfMonth);

        etMoney.setText(String.valueOf(recordToUpdate.getMoney()));
        etDetail.setText(recordToUpdate.getDetail());

        List<Type> types = TypeDao.getTypesByType(recordToUpdate.getType());
        type = recordToUpdate.getType();
        rgType.check(type == TypeEnum.IN.value() ? R.id.rb_type_in : R.id.rb_type_out);
        setSpinnerContent();

        int selected = 0;
        int size = types.size();
        for (int i = 0; i < size; i++) {
            if (types.get(i).getName().equals(recordToUpdate.getTypeName())) {
                selected = i;
                break;
            }
        }
        spinnerType.setSelection(selected, true);

    }

    @Override
    public void onResume() {
        super.onResume();
        setSpinnerContent();
        if (isUpdate) {
            initData();
        }
    }

    private boolean saveOrUpdateRecord() {
        int y = datePicker.getYear();
        int m = datePicker.getMonth() + 1;
        int d = datePicker.getDayOfMonth();
        String date = DateUtil.getFormatYMD(y, m, d);
        return RecordDao.addOrUpdateRecord(isUpdate,
                recordToUpdate.getId(),
                date,
                Double.parseDouble(etMoney.getText().toString()),
                etDetail.getText().toString(),
                type,
                (String) spinnerType.getSelectedItem());
    }

    /**
     * 根据收支的选择填充spinner的内容
     */
    private void setSpinnerContent() {
        List<Type> types = TypeDao.getTypesByType(type);
        if (types == null || types.size() == 0) {
            ToastUtils.showShort(R.string.add_no_type);
            return;
        }
        List<String> typeNames = new ArrayList<>();
        for (Type t : types) {
            typeNames.add(t.getName());
        }
        ArrayAdapter spinnerAdapter = new ArrayAdapter<>(getContext(), R.layout.item_spinner, typeNames);
        spinnerAdapter.setDropDownViewResource(R.layout.item_spinner_dropdown);
        spinnerType.setAdapter(spinnerAdapter);
    }

    /**
     * 重置清空
     */
    private void reset() {
        Calendar c = Calendar.getInstance();
        datePicker.updateDate(c.get(Calendar.YEAR), c.get(Calendar.MONTH), c.get(Calendar.DATE));
        etMoney.setText("");
        etDetail.setText("");
        type = TypeEnum.OUT.value();
        rgType.check(R.id.rb_type_out);
        setSpinnerContent();
    }

}

File: app/src/main/java/com/ddancn/keepaccount/fragment/AddFragment.kt
package com.ddancn.keepaccount.fragment

import android.annotation.SuppressLint
import android.content.Context
import androidx.annotation.NonNull
import com.ddancn.keepaccount.R
import com.ddancn.keepaccount.activity.SettingActivity
import com.ddancn.keepaccount.activity.UpdateActivity
import com.ddancn.keepaccount.adapter.AddCategoryAdapter
import com.ddancn.keepaccount.constant.TypeEnum
import com.ddancn.keepaccount.dao.CategoryDao
import com.ddancn.keepaccount.dao.RecordDao
import com.ddancn.keepaccount.databinding.FragmentAddBinding
import com.ddancn.keepaccount.entity.Category
import com.ddancn.keepaccount.entity.Record
import com.ddancn.keepaccount.util.getFormatYMD
import com.ddancn.lib.base.BaseFragment
import com.google.android.flexbox.AlignItems
import com.google.android.flexbox.FlexboxLayoutManager
import com.google.android.flexbox.JustifyContent
import java.util.*

/**
 * @author ddan.zhuang
 * 添加或修改记录页面
 */
class AddFragment : BaseFragment<FragmentAddBinding>() {

    private var type = TypeEnum.OUT.value()
    private var isUpdate = false
    private lateinit var recordToUpdate: Record

    private val categoryAdapter = AddCategoryAdapter(R.layout.item_add_category)
    private var selectedCate: Category? = null

    override fun hasHeader(): Boolean = !isUpdate

    override fun onAttach(@NonNull context: Context) {
        super.onAttach(context)
        isUpdate = context is UpdateActivity
    }

    override fun onStart() {
        super.onStart()
        recordToUpdate = (activity as? UpdateActivity)?.recordToUpdate ?: Record()
//        view?.postDelayed({
//            KeyboardUtils.showSoftInput(vb.etMoney)
//        }, 500)
    }

    override fun initView() {
        headerView.setTitle(R.string.app_name)
        headerView.setRightImage(R.drawable.ic_setting)

        val manager = FlexboxLayoutManager(context)
        // 主轴方向排列
        manager.justifyContent = JustifyContent.FLEX_START
        // 次轴方向排列
        manager.alignItems = AlignItems.CENTER
        vb.rvCategory.layoutManager = manager
        vb.rvCategory.adapter = categoryAdapter
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun bindListener() {
        headerView.setRightClickListener { start(SettingActivity::class.java) }
        // 选择收支大类，联动改变类型列表的内容
        vb.rgType.setOnCheckedChangeListener { group, checkedId ->
            type = if (checkedId == R.id.rb_type_out) TypeEnum.OUT.value() else TypeEnum.IN.value()
            setCategoryList()
        }
        categoryAdapter.setOnItemClickListener { adapter, view, position ->
            selectedCate = categoryAdapter.data[position]
            categoryAdapter.setSelectedPos(position)
        }
        // 按钮点击事件
        vb.btnAdd.setOnClickListener {
            if (!checkParamLegal()) {
                return@setOnClickListener
            }
            if (saveOrUpdateRecord()) {
                toast(if (isUpdate) R.string.update_succeed else R.string.add_succeed)
                if (isUpdate) {
                    activity?.finish()
                } else {
                    reset()
                }
            } else {
                toast(if (isUpdate) R.string.update_fail else R.string.add_fail)
            }
        }
    }

    /**
     * 判断输入是否合法
     *
     * @return 是否合法
     */
    private fun checkParamLegal(): Boolean {
        if (vb.etMoney.text.isNullOrBlank()) {
            toast(R.string.add_need_money)
            return false
        }
        if (selectedCate == null) {
            toast(R.string.add_need_category)
            return false
        }
        return true
    }

    /**
     * 如果是修改页面，先填上原来的数据
     */
    private fun fillOldData() {
        vb.btnAdd.text = getString(R.string.update_btn)
        // 选择日期
        recordToUpdate.date.split("-").let {
            val year = it[0].toInt()
            val month = it[1].toInt()
            val dayOfMonth = it[2].toInt()
            vb.datePicker.updateDate(year, month - 1, dayOfMonth)
        }
        vb.etMoney.setText(recordToUpdate.money.toString())
        vb.etDetail.setText(recordToUpdate.detail)

        // 填充收支类型判断
        type = recordToUpdate.type
        val categories = CategoryDao.getCategoriesByType(type)
        vb.rgType.check(if (type == TypeEnum.IN.value()) R.id.rb_type_in else R.id.rb_type_out)
        setCategoryList()

        // 填充具体类型名称
        var selected = 0
        categories.forEachIndexed { index, category ->
            if (category.name == recordToUpdate.categoryName) {
                selected = index
            }
        }
        categoryAdapter.setSelectedPos(selected)
        selectedCate = categoryAdapter.data[selected]
    }

    override fun onResume() {
        super.onResume()
        if (isUpdate) {
            fillOldData()
        } else {
            setCategoryList()
        }
    }

    /**
     * 添加或修改记录
     */
    private fun saveOrUpdateRecord(): Boolean {
        val y = vb.datePicker.year
        val m = vb.datePicker.month + 1
        val d = vb.datePicker.dayOfMonth
        val date = getFormatYMD(y, m, d)
        val record = Record(
            date = date,
            money = vb.etMoney.text.toString().toDouble(),
            detail = vb.etDetail.text.toString(),
            type = type,
            categoryName = selectedCate!!.name
        )

        return if (isUpdate) {
            record.id = recordToUpdate.id
            RecordDao.updateRecord(record)
        } else {
            RecordDao.addRecord(record)
        }
    }

    /**
     * 根据收支的选择填充类型的内容
     */
    private fun setCategoryList() {
        val categories = CategoryDao.getCategoriesByType(type)
        if (categories.isEmpty()) {
            toast(R.string.add_no_category)
            categoryAdapter.setNewData(emptyList())
            selectedCate = null
            return
        }

        categoryAdapter.setNewData(categories)
        categoryAdapter.setSelectedPos(0)
        selectedCate = categoryAdapter.data[0]
    }

    /**
     * 重置清空
     */
    private fun reset() {
        val calendar = Calendar.getInstance()
        vb.datePicker.updateDate(
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DATE)
        )
        vb.etMoney.setText("")
        vb.etDetail.setText("")
        type = TypeEnum.OUT.value()
        vb.rgType.check(R.id.rb_type_out)
        setCategoryList()
    }

}
